server/chat-plugins/teams.ts

```diff
@@ -0,0 +1,686 @@
+			`SELECT * FROM teams WHERE teamid = $1`, [teamid],
+		);
+		if (!rows.length) return null;
+		return rows[0] as StoredTeam;
+	}
+	async delete(id: number) {
+		await this.query(
+			`DELETE FROM teams WHERE teamid = $1`, [id],
+		);
+	}
+};
+
+export const destroy = () => TeamsHandler.destroy();
+
+export const commands: Chat.ChatCommands = {
+	teams: {
+		upload() {
+			return this.parse('/j view-teams-upload');
+		},
+		update: 'save',
+		async save(target, room, user, connection, cmd) {
+			TeamsHandler.validateAccess(connection, true);
+			const targets = Utils.splitFirst(target, ',', 5);
+			const isEdit = cmd === 'update';
+			const rawID = isEdit ? targets.shift() : undefined;
+			const teamID = Number(rawID || "");
+			let [teamName, formatid, rawPrivacy, rawTeam] = targets;
+			if (isEdit && isNaN(teamID)) {
+				connection.popup("Invalid team ID provided.");
+				return null;
+			}
+			if (rawTeam.includes('\n')) {
+				rawTeam = Teams.pack(Teams.import(rawTeam));
+			}
+			if (!rawTeam) {
+				connection.popup("Invalid team.");
+				return null;
+			}
+			formatid = toID(formatid);
+			teamName = toID(teamName) ? teamName : null!;
+			const id = await TeamsHandler.save(
+				connection, formatid, rawTeam, teamName, !!Number(rawPrivacy), isEdit ? teamID : undefined
+			);
+
+			const page = isEdit ? 'edit' : 'upload';
+			if (id) {
+				connection.send(`>view-teams-${page}\n|deinit`);
+				this.parse(`/join view-teams-view-${id}-${id}`);
+			} else {
+				this.parse(`/join view-teams-${page}`);
+				return;
+			}
+		},
+		''(target) {
+			return this.parse('/teams user ' + toID(target) || this.user.id);
+		},
+		latest() {
+			return this.parse(`/j view-teams-filtered-latest`);
+		},
+		views: 'mostviews',
+		mostviews() {
+			return this.parse(`/j view-teams-filtered-views`);
+		},
+		user: 'view',
+		for: 'view',
+		view(target) {
+			const [name, rawNum] = target.split(',').map(toID);
+			const num = parseInt(rawNum);
+			if (rawNum && isNaN(num)) {
+				return this.popupReply(`Invalid count.`);
+			}
+			let page = 'view';
+			switch (this.cmd) {
+			case 'for': case 'user':
+				page = 'all';
+				break;
+			}
+			return this.parse(`/j view-teams-${page}-${toID(name)}${num ? `-${num}` : ''}`);
+		},
+		async delete(target, room, user, connection) {
+			TeamsHandler.validateAccess(connection, true);
+			const teamid = parseInt(target);
+			if (isNaN(teamid)) return this.popupReply(`Invalid team ID.`);
+			const teamData = await TeamsHandler.get(teamid);
+			if (!teamData) return this.popupReply(`Team not found.`);
+			if (teamData.ownerid !== user.id && !user.can('rangeban')) {
+				return this.errorReply("You cannot delete teams you do not own.");
+			}
+			await TeamsHandler.delete(teamid);
+			this.popupReply(`Team ${teamid} deleted.`);
+			for (const page of connection.openPages || new Set()) {
+				if (page.startsWith('teams-')) this.refreshPage(page);
+			}
+		},
+		async setprivacy(target, room, user, connection) {
```

Comment:
There should be a third level of privacy: unlisted

Currently, your only two options are to either have it completely open or only viewable by yourself. This doesn't support a use case where you built a team for a teammate but obviously don't want it public but want your teammate. Unlisted would basically be a team link but with an extra `password` field that's randomly generated for a team a la private PS replays.

Reply:
Not exactly. Right now, we have 'searchable' or 'unlisted' is probably a more accurate way of putting it. Anyone can view a team if they have the number, but it'll only show up in searches if it's marked as public.

Reply:
Now that I say this to myself, team IDs should probably be hashes so they aren't guessed easily.